src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/BinaryMessageParser.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Binary;
+using System.Buffers;
+
+namespace Microsoft.AspNetCore.Sockets.Internal.Formatters
+{
+    internal class BinaryMessageParser
+    {
+        private ParserState _state;
+
+        public void Reset()
+        {
+            _state = default(ParserState);
+        }
+
+        public bool TryParseMessage(ref BytesReader buffer, out Message message)
+        {
+            if (_state.Length == null)
+            {
+                var length = buffer.TryReadBytes(sizeof(long))?.ToSingleSpan();
+                if (length == null || length.Value.Length < sizeof(long))
+                {
+                    message = default(Message);
+                    return false;
+                }
+
+                var longLength = length.Value.ReadBigEndian<long>();
+                if (longLength > Int32.MaxValue)
+                {
+                    throw new FormatException("Messages over 2GB in size are not supported");
+                }
+                buffer.Advance(length.Value.Length);
+                _state.Length = (int)longLength;
+            }
+
+            if (_state.MessageType == null)
+            {
+                if (buffer.Unread.Length == 0)
+                {
+                    message = default(Message);
+                    return false;
+                }
+
+                var typeByte = buffer.Unread[0];
+
+                if (!TryParseType(typeByte, out var messageType))
+                {
+                    throw new FormatException($"Unknown type value: 0x{typeByte:X}");
+                }
+
+                buffer.Advance(1);
+                _state.MessageType = messageType;
+            }
+
+            if (_state.Payload == null)
+            {
+                _state.Payload = new byte[_state.Length.Value];
+            }
+
+            while (_state.Read < _state.Payload.Length && buffer.Unread.Length > 0)
+            {
+                // Copy what we can from the current unread segment
+                var toCopy = Math.Min(_state.Payload.Length - _state.Read, buffer.Unread.Length);
+                buffer.Unread.Slice(0, toCopy).CopyTo(_state.Payload.Slice(_state.Read));
+                _state.Read += toCopy;
+                buffer.Advance(toCopy);
+            }
+
+            if (_state.Read >= _state.Payload.Length)
```

Comment:
Only if my math is wrong. I can put an assert in, or just swap it to `==`. I had `==` originally but was being paranoid :).